Feature/branches coverage support

som4ik opened the pull request.

gleseur committed:
679 Stricter coverage percentage computation to ensure 100% coverage means exactly 100%

hedgesky commented on lib/simplecov/combiners/base_combiner.rb:
```diff
@@ -0,0 +1,34 @@
+      end
+
+      def combine
+        puts "override this method"
```
FYI: there is `NotImplementedError` error in Ruby for similar cases.

hedgesky commented on lib/simplecov.rb:
```diff
@@ -67,25 +68,63 @@ def add_not_loaded_files(result)
+    # @return [Hash]
+    #
+    def result_with_not_loaded_files
+      @result = SimpleCov::Result.new add_not_loaded_files(@result)
```
https://github.com/rubocop-hq/ruby-style-guide#method-invocation-parens

hedgesky commented on lib/simplecov.rb:
```diff
@@ -67,25 +68,63 @@ def add_not_loaded_files(result)
     end
 
+    #
+    # Unite the result not matter what coverage type called
```
"Unite the result so it wouldn't matter what coverage type was called"

hedgesky commented on lib/simplecov.rb:
```diff
@@ -67,25 +68,63 @@ def add_not_loaded_files(result)
+
+    #
+    # Filter coverage result
+    # The result before filter also have result of coverage for files
```
"also *has*"

som4ik commented:
Add branch coverage measurement statistics (ruby ~> 2.5) 
In this PR:
- Add configuration needed to trigger this new functionality.
- Unite the results no matter what coverage version used to be there.
- Add `branch` class and related support methods to `SourceFile`
- Replace `RawCoverage` with `RunResultsCombiner`, which breaks logic of combining different results in to file,line and branch combining level, this new structor adds more readability and makes it open for modification (ex: adding methods coverage or any other coverage in future)
- Remove useless files from the coverage result before starting any operations on them as `Coverage.start` take in case all loaded files including required gems which files are not inside the project directory
- On simplecov-html side we will have this visual statistics:
![branch1](https://user-images.githubusercontent.com/7699543/43396656-e273be44-940a-11e8-99b3-053929752e3d.png)
 `[0, '-'] => [passing count, branch sign]`

hedgesky commented on lib/simplecov.rb:
```diff
@@ -67,25 +68,63 @@ def add_not_loaded_files(result)
+    #
+    # @return [Hash]
+    #
+    def adapt_coverage_result
```
Should methods `adapt_coverage_result`, `remove_useless_results`, `result_with_not_loaded_files` and `process_coverage_result` belong to public interface?

som4ik replied:
Totally agree, they should be defined as private methods.

hedgesky commented on lib/simplecov/combiners/base_combiner.rb:
```diff
@@ -0,0 +1,32 @@
+module SimpleCov
+  module Combiners
+    #
+    # Represents the base behavior or the combiner
```
Typo: `or` instead of `of`

som4ik replied:
fixed

hedgesky commented on lib/simplecov/combiners/base_combiner.rb:
```diff
@@ -0,0 +1,32 @@
+  module Combiners
+    #
+    # Represents the base behavior or the combiner
+    # take two coverage statisitcs and combine them depends on the logic needed
```
Typo: `statisitcs`. Also, `depending` instead of `depends` should be used here.

som4ik replied:
fixed

hedgesky commented on lib/simplecov/combiners/branches_combiner.rb:
```diff
@@ -0,0 +1,42 @@
+    #
+    class BranchesCombiner < BaseCombiner
+      #
+      # Return merged branches or the existed branche if other is missing
```
Typo: `branche`

som4ik replied:
fixed

hedgesky commented on lib/simplecov/combiners/branches_combiner.rb:
```diff
@@ -0,0 +1,42 @@
+
+      #
+      # Branches inside files are always same if they exists, the difference only in coverage count
+      # Branch coverage report for any conditional case is build from hash it's key is condition and
```
`is built` vs `is build`. Also, a comma or a dot should be place before `it's key` because those are different sentences.

som4ik replied:
fixed

hedgesky commented on lib/simplecov/combiners/branches_combiner.rb:
```diff
@@ -0,0 +1,42 @@
+      # Branch coverage report for any conditional case is build from hash it's key is condition and
+      # body also hash with hashes inside << keys from condition and value is coverage rate >>
+      # ex: branches =>{ [:if, 3, 8, 6, 8, 36] => {[:then, 4, 8, 6, 8, 12] => 1, [:else, 5, 8, 6, 8, 36]=>2}, other branches...}
+      # So we create copy of result and update it values depends on the combined branches coverage values
```
`depending` over `depends`

som4ik replied:
fixed

hedgesky commented on lib/simplecov/combiners/branches_combiner.rb:
```diff
@@ -0,0 +1,42 @@
+      # @return [Hash]
+      #
+      def combine_branches
+        combine_result = first_coverage.clone
```
`combine_result` sounds like an action, but it's a hash. Please use `combined_result` as a variable name.

som4ik replied:
Done.

hedgesky commented on lib/simplecov/combiners/files_combiner.rb:
```diff
@@ -0,0 +1,67 @@
+      def combine
+        return existed_coverage unless empty_coverage?
+
+        combine_lines_coverage
```
Why do we need blank lines here?

som4ik replied:
For better readability.

hedgesky commented on lib/simplecov/combiners/files_combiner.rb:
```diff
@@ -0,0 +1,67 @@
+      # @return [Hash]
+      #
+      def combine_lines_coverage
+        combined_results[:lines] = SimpleCov::Combiners::LinesCombiner.combine!(
```
Why do we specify full path to the module? The `SimpleCov::Combiners` part could be omitted, resulting in shorter and more readable line. Below — the same.

hedgesky commented on lib/simplecov/configuration.rb:
```diff
@@ -71,6 +71,14 @@ def filters
     end
 
+    #
+    # Details covers all measurement types of only lines
```
"Details covers all measurement types of only lines" — not clear what it should mean. Also, absence of the dot and starting next line with small letter is confusing.

som4ik replied:
Fixed.

hedgesky commented on lib/simplecov/result_adapter.rb:
```diff
@@ -0,0 +1,31 @@
+
+module SimpleCov
+  #
+  # Resposible of adapting the format of the coverage result weather it's default or with statistics
```
`Resposible` — typo. `weather` — typo.

som4ik replied:
Fixed.

hedgesky commented on lib/simplecov/result_adapter.rb:
```diff
@@ -0,0 +1,31 @@
+    end
+
+    def adapt
+      return result unless result
```
Maybe we can replace it with `return unless result`?

som4ik replied:
Yes , I agree.

hedgesky commented on lib/simplecov/run_file_coverage.rb:
```diff
@@ -0,0 +1,38 @@
+
+module SimpleCov
+  #
+  # Get certain file coverage report by apply
```
Please reword the comment.

som4ik replied:
Done.

hedgesky commented on lib/simplecov/run_file_coverage.rb:
```diff
@@ -0,0 +1,38 @@
+  private
+
+    attr_reader :lines, :absolute_path
+    attr_accessor :classified_result
```
This accessor is never used.

som4ik replied:
Removed.

hedgesky commented on lib/simplecov/run_results_combiner.rb:
```diff
@@ -0,0 +1,65 @@
+  # There might be reports from different kinds of tests,
+  # e.g. RSpec and Cucumber. We need to combine their results
+  # into unified one. This class does that.
+  # To unite results on per file basis, it leverages
```
part of comment is missed after "it leverages"

som4ik replied:
Fixed.

hedgesky commented on lib/simplecov/run_results_combiner.rb:
```diff
@@ -0,0 +1,65 @@
+    end
+
+    #
+    # Combine process explination
```
`explination` — typo

som4ik replied:
Fixed.

hedgesky commented on lib/simplecov/source_file.rb:
```diff
@@ -72,6 +73,98 @@ def status
 
+    #
+    # Representing single branch that been detected in coverage report
+    # Give us support methods that handle neede calculations
```
`neede` — typo

som4ik replied:
Fixed

hedgesky commented on spec/source_file_branch_spec.rb:
```diff
@@ -0,0 +1,152 @@
+
+if SimpleCov.usable?
+  describe SimpleCov::SourceFile::Branch do
+    let(:resutls) do
```
`resutls` — typo

som4ik replied:
Fixed.

hedgesky commented on spec/source_file_branch_spec.rb:
```diff
@@ -0,0 +1,152 @@
+# frozen_string_literal: true
```
Currently I have no time to review tests carefully. Will return to them later

hedgesky commented on lib/simplecov/result_adapter.rb:
```diff
@@ -0,0 +1,30 @@
+
+      result.each_with_object({}) do |(file_name, cover_statistic), adapted_result|
+        if cover_statistic.is_a?(Array)
+          adapted_result.merge!(file_name => {:lines => cover_statistic})
```
`lines: ` notation could be used here

som4ik replied:
fixed

som4ik replied:
Rubocop specify this modification as offense.  
It requesting to use `hash rockets syntax`.

hedgesky commented on lib/simplecov/source_file.rb:
```diff
@@ -72,6 +73,98 @@ def status
     end
 
+    #
+    # Representing single branch that been detected in coverage report
```
`has been`. Dot at the end of the sentence.

som4ik replied:
fixed

hedgesky commented on lib/simplecov/source_file.rb:
```diff
@@ -72,6 +73,98 @@ def status
+      attr_accessor :coverage, :root_id
+
+      def initialize(*branch_attrs, root_id)
+        @type       = branch_attrs[0]
```
We can use splat assignment here

deivid-rodriguez committed:
Fix cucumber deprecations

```
Deprecated: Found tags option '~@wip'. Support for '~@tag' will be removed from the next release of Cucumber. Please use 'not @tag' instead.
Deprecated: Found tags option '~@exclude-ruby'. Support for '~@tag' will be removed from the next release of Cucumber. Please use 'not @tag' instead
```

deivid-rodriguez committed:
Drop phantomjs + poltergeist in favor of apparition

deivid-rodriguez committed:
Bump capybara to "~> 3.29"

andyw8 committed:
Add an alternative way to work with Spring (#732)

Enabling eager loading will mean every test takes additional time to run.
Instead, we can disable Spring using an environment variable.

PragTob committed:
Create Code of Conduct (#734)

It's way past time, only an oversight that we didn't have this before :)

PragTob committed:
Point to CoC in README

tchia04 committed:
Fix the Coverage Link

colszowka committed:
Merge pull request #735 from tchia04/fix_broken_link_to_coverage

Fix the Coverage Link

PragTob committed:
Use JDK8 to avoid JDK11 problems/failures

https://github.com/jruby/jruby/issues/5736

belfazt committed:
Warn the user if the minimum coverage is set greater than 100 (#737)

* Warn the user if the minimum coverage that is being set is greater than 100

ohboyd committed:
Clarify browser startup process by adding bullet

ohboyd committed:
Add consistent line lengths for readability

harpocrates12 committed:
Add new line to exit_status log. (#742)

ohboyd committed:
Note mac-specific ex. and link to commands for other operating systems

Commit:
Fix syntax highlighting in readme

annaswims committed:
Avoid hanging with parallel_tests

ParallelTests.last_process? retuns true if the process is the last one to start.  We are
interested in the last test to finish so we want to use ParallelTests.number_of_running_processes

Because it's possible to have the parallel_tests gem in your project and still run tests serially we
want to also check if tests are currently being run in parallel.  If ENV['TEST_ENV_NUMBER'] is defined
we know tests are currently being run in parallel.

PragTob committed:
Changelog for important parallel tests hanging bugfix

JacobEvelyn committed:
Add option to disable error status code printing (#747)

PR #688 added functionality that broke the
[`JacobEvelyn/friends`](https://github.com/JacobEvelyn/friends)
build (see https://travis-ci.com/JacobEvelyn/friends/jobs/238680478),
which runs many CLI processes and parses the output. Some CLI
processes are expected to have non-success status codes, but
these tests also check that our CLI wrote the correct message
to STDERR, and with #688 the output to STDERR is now changed.

This commit adds a SimpleCov configuration flag to disable the
message added in #688.

thomas07vt committed:
Update cucumber to 3.x (#748)

This removes the Transform deprecation warning

PragTob committed:
Update version on dev to indicate current state

Seemed to be confusing/annoying for some people #752

Commit:
Fix/rubocop update (#754)

* updated cop

* corrected files according to new cop

adam12 committed:
Update references to Minitest constant (#757)

hmtanbir committed:
add debian/ubuntu terminal cmd in README (#753)

* add linux terminal cmd in README

* update readme

change linux to ubuntu

PragTob committed:
ParameterType seems to be cucumber 3.x only (#759)

PragTob committed:
test against newest JRuby (#761)

PragTob committed:
test against ruby 2.7 preview but also allow it to fail (#760)

Commit:
Add Simplecov-Material Formatter availabilty (#763)

tedt10e committed:
Update Note for JRuby User (#765)

deivid-rodriguez committed:
Bump rubocop to 0.53.0

Since that version introduced `Style/TrailingCommaInArrayLiteral` and
`Style/TrailingCommaInHashLiteral`, which we are using.

Fix/ignore the new offenses introduced in the new version, and fix new
deprecations in the configuration.

deivid-rodriguez committed:
Update configuration to match the style we're using

MSP-Greg committed:
README.md - update 'Getting Started' [skip ci]

deivid-rodriguez committed:
Bump CI things (#768)

* Update rubygems in a simpler way

* Use latest rubygems

* Use the latest patch level rubies

* Use whatever rubygems comes with ruby

deivid-rodriguez committed:
Improve development setup (#770)

* Remove unneeded gitignores

These files no longer exist.

* Commit lock file

* Centralize development dependencies in Gemfile

This is more flexible, since you can point to gems from git or path
sources, for example.

PragTob committed:
tidy up chanelog

PragTob committed:
Clarify interplay with common view libraries

som4ik committed:
add branch coverage feature

som4ik committed:
fix rubocop offences after rebase

som4ik committed:
fix CR notices and add ability for getting results from parallel tests :boom:

som4ik committed:
Fix CR notices and do some report improvements
